substrate/client/service/src/builder.rs

```diff
@@ -777,65 +781,63 @@ where
 }
 
 /// Parameters to pass into `build_network`.
-pub struct BuildNetworkParams<
-	'a,
-	TBl: BlockT,
-	TNet: NetworkBackend<TBl, <TBl as BlockT>::Hash>,
-	TExPool,
-	TImpQu,
-	TCl,
-> {
+pub struct BuildNetworkParams<'a, Block, Net, TxPool, IQ, Client>
+where
+	Block: BlockT,
+	Net: NetworkBackend<Block, <Block as BlockT>::Hash>,
+{
 	/// The service configuration.
 	pub config: &'a Configuration,
 	/// Full network configuration.
-	pub net_config: FullNetworkConfiguration<TBl, <TBl as BlockT>::Hash, TNet>,
+	pub net_config: FullNetworkConfiguration<Block, <Block as BlockT>::Hash, Net>,
 	/// A shared client returned by `new_full_parts`.
-	pub client: Arc<TCl>,
+	pub client: Arc<Client>,
 	/// A shared transaction pool.
-	pub transaction_pool: Arc<TExPool>,
+	pub transaction_pool: Arc<TxPool>,
 	/// A handle for spawning tasks.
 	pub spawn_handle: SpawnTaskHandle,
 	/// An import queue.
-	pub import_queue: TImpQu,
+	pub import_queue: IQ,
 	/// A block announce validator builder.
-	pub block_announce_validator_builder:
-		Option<Box<dyn FnOnce(Arc<TCl>) -> Box<dyn BlockAnnounceValidator<TBl> + Send> + Send>>,
-	/// Optional warp sync config.
-	pub warp_sync_config: Option<WarpSyncConfig<TBl>>,
+	pub block_announce_validator_builder: Option<
+		Box<dyn FnOnce(Arc<Client>) -> Box<dyn BlockAnnounceValidator<Block> + Send> + Send>,
+	>,
+	/// Syncing strategy to use in syncing engine.
+	pub syncing_strategy: Box<dyn SyncingStrategy<Block>>,
 	/// User specified block relay params. If not specified, the default
 	/// block request handler will be used.
-	pub block_relay: Option<BlockRelayParams<TBl, TNet>>,
+	pub block_relay: Option<BlockRelayParams<Block, Net>>,
 	/// Metrics.
 	pub metrics: NotificationMetrics,
 }
 
 /// Build the network service, the network status sinks and an RPC sender.
-pub fn build_network<TBl, TNet, TExPool, TImpQu, TCl>(
-	params: BuildNetworkParams<TBl, TNet, TExPool, TImpQu, TCl>,
+pub fn build_network<Block, Net, TxPool, IQ, Client>(
```

Comment:
While I'm late to the party (sorry).

Before this goes into any stable release or whatever, can we please revert this? 

Basically we should not expose `build_polkadot_syncing_strategy` . Instead we should do this internally in this `build_network` function and then expose some `build_network_advanced` or whatever for @nazar-pc that takes then the syncing strategy. This way there is no breaking change to the outside for developers. (that only have a node and not any custom syncing stuff)

The end goal should be that all the `node/service` changes to all the nodes in this repo are reverted.

@lexnv or @dmitry-markin can you please take care of this?

Reply:
https://github.com/paritytech/polkadot-sdk/pull/5737 actually extracted the whole syncing strategy out of `build_network` and exposed `build_default_syncing_engine` function to construct the typical setup.

The fact that `build_network` was not just building the network, but also syncing engine, syncing strategies and hardcoding all the protocol was a long standing problem IMO. It was also taking (and still does) the whole `sc_service::config::Configuration`, most of which it doesn't use, which is also a problem (though I didn't address it in any of my changes). It makes using Substrate as a library quite painful for those who need to customize it the way we do at Subspace.

My goal would be for `build_network` to be simplified to little more than `Net::new()` call. The fact that it unconditionally adds a whole bunch of the protocols is annoying and undesirable.

Having `build_network_advanced` is an option, but not a long-term solution the way I imagine it. Happy to push it into https://github.com/paritytech/polkadot-sdk/pull/5737 if that is something you feel strongly about.

Reply:
> I guess the gist of it is that `build_network` should build the network, not syncing engines and other stuff it is coupled with right now.

I mean I'm not against this and I support this. Just about how it is done. I mean just keep the function do what it was doing all the time and then provide some extended way.

Reply:
Reverted `build_network`'s API to before this PR and added `build_network_advanced` instead in https://github.com/paritytech/polkadot-sdk/pull/5737